design-proposals/klip-15-new-api-and-client.md

```diff
@@ -0,0 +1,187 @@
+* The implementation will support equivalents for all the operations that are currently available on the HTTP/REST API
+* The implementation will be designed in a reactive / non blocking way in order to provide the best performance / scalability characteristics with
+low resource usage. This will also influence the overall threading model of the server and position us with a better, more scalability internal
+server architecture that will help future proof the ksqlDB server.
+* The problematic parts of the current HTTP API will be removed (query streaming using chunked response and websockets)
+* The remainder of the current HTTP API will remain supported as-is using the new implementation.
+
+## What is not in scope
+
+* We are not creating a new RESTful API.
+* We are not currently looking to write clients for languages other than Java and JavaScript (although those may follow later)
+* We will not initially allow for consuming directly from Kafka topics using the client (although tha may follow later)
+
+## Value/Return
+
+We hope that by providing a delightful, easy to use client, it will enable application developers to easily write powerful
+applications that take advantage of their data plane / event streaming platform more effectively.
+
+We hope that this could be transformative for the project in terms of adoption as it would position ksqlDB as a great choice for
+writing typical event sourcing / CQRS / DDD style applications, which are currently hard to write using ksqlDB alone. 
+
+There are also further incidental advantages gained by this work. By using a modern server side implementation such as Vert.x 
+there are benefits in relation to performance, scalability, simplicity of implementation, 
+and threading model. Not to mention reduced dependencies and better resource usage. This will set us up better for the kinds of high
+throughput operations that we will need to implement efficiently now that the project has pivoted to a more active application
+facing server rather than a more passive DDL engine.
+
+## Public APIS
+
+The following changes / additions to APIs will occur:
+
+* The current chunked streaming and websockets streaming endpoints will be retired.
+* The client will provide a new public API
+* We will provide a new wire protocol for client<->server communication.
+
+## Design
+
+### The client
+
+* The Java client will provide an API based on JDK 9 flow / reactive streams and completable future
+* The networking will be handled by Vert.x (which uses Netty). Both Vert.x and Netty are very widely
+used toolkits which are known for great performance, and have few dependencies.
+* The client will have minimal transitive jar dependencies - this is important as the client will be embedded in end user applications.
+* Client connections are designed to be re-used.
+* The client will be thread-safe.
+
+### The server
+
+* The toolkit used on the server side will be Vert.x
+* The current Jetty / JAX-RS usage will be retired.
```

Comment:
> Despite being genuinely excited to learn Vert.x and rewrite our bulky server, I'm wondering if we can de-risk this KLIP by splitting out the server rewrite from the client/protocol work. Is there any way our existing server implementation can be used for phase 1? That way we can move forward with the wire protocol and the client implementation without working on a server 2.0.

Yes, absolutely - I think the work on hosting the new protocol, and porting the existing API to Vert.x can be done in different phases. It just means we would have both Vert.x and Jetty servers running in parallel until the old API is ported.

Reply:
> The reason is that we need to make sure that all of the existing functionality that depends on the confluent's rest-utils (some port of dropwizard) and the jakarta/hk2 injection mechanisms still work as these are the standard method for confluent components (and there are plugins that already integrate with these injection mechanisms). Perhaps it is possible with Vert.x (and it's not unlikely that I'm totally missing the
> 
> @rodesai / @spena might be able to shine some light on this

I chatted with @spena a couple of days about this and Sergio kindly showed me the code for the current Jetty plugins. It seemed to me they could be fairly easily ported to Vert.x, although there may be some angles I haven't completely understood.

Reply:
I agree about decoupling the API design versus the threading model from this KLIP.

Here's a java client lib that I wrote against the current API : https://github.com/daniellavoie/ksqldb-java-client

It's fully reactive (RxJava like) with Project Reactor. It supports Observable through Mono and Flux. I currently support all the push and pull queries in a observable way. The same library could be easily written for Node.js server but also frontend Javascript and Typescript.

From a developer perspective, having that client API is a good start. How the server behaves from performance perspective is not my problem. Reactive or blocking, I shouldn't be aware of that. I'm all for reactive. I would recommend Webflux from Spring over VertX but that is a different discussion. Simple HTTP 1.1 with  REST and SSE events (that are super well supported on the browser) are good enough from a client perspective. gRPC, WebSocket and anything else is just friction and more complication for the developer. If the backend server is reactive, SSE will not exhaust the thread pool from the server.

Reply:
Hey @daniellavoie 
I think it's important that our client API isn't tightly coupled to Spring, which is why I use vanilla jdk 9 flow interfaces (ok, they are backported for now!). (Not everyone wants to use Spring).
For those that want a very Spring Reactor style API, it would be a simple matter to provide a shim over our client to expose Mono/Flux etc :)

Reply:
@purplefox Reactor as no dependencies on Spring. It's a foundation for Spring. So clients pulls no dependencies from it. But I agree with you that offering a shim would be better approach. I'm concerned about the JDK 9 flow interface because it would enforce clients to use JDK11+. A lot of enterprise will be running Java 8 for a long time. Reactive Streams would probably the best interface for maximum compatibility (Flow is based on it actually).

Reply:
Apologies - I use "Spring" in a loose sense - meaning the Spring ecosystem, and reactor is a part of that. I should have said "I think it's important that our client API isn't tightly coupled to Project Reactor". Flow is in JDK 9 (not 11), but I agree, better standardise on Reactive Streams and optionally provide a JDK 9 flow shim for those who are using Java 9+

Reply:
I mentionned JDK 11 because 9 and 10 are not LTS. Only 11.